core/trino-main/src/main/java/io/trino/operator/TopNOperator.java

```diff
@@ -190,12 +205,23 @@ public void addInput(Page page)
     @Override
     public void finish()
     {
+        isPartialFlushing = false;
         pageBuffer.finish();
     }
 
+    private boolean isBuilderFull()
+    {
+        return maxPartialMemory <= topNProcessor.getEstimatedSizeInBytes();
+    }
+
     private void addPage(Page page)
     {
+        checkState(!isBuilderFull(), "TopN buffer is already full");
         topNProcessor.addInput(page);
+        if (partial && isBuilderFull()) {
+            isPartialFlushing = true;
+            pageBuffer.add(new Page(1)); // create a dummy page in order to call TopNPages.process()
```

Comment:
@sopel39 this looks weird, what is the right way to do this?

Reply:
I don't like this approach, either. This is like picking the lesser evil. @sopel39 if you have a better way to accomplish this, please let me know.

Reply:
I changed the implementation to add one more method in Transformation interface. With this, operators implementing WorkProcessor can handle input buffer full situation.